src/editor/codemirror/dnd.ts

```diff
@@ -0,0 +1,77 @@
+import { ChangeSet, Transaction } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { pythonSnippetMediaType } from "../../common/mediaTypes";
+import { calculateChanges } from "./edits";
+
+// No access to the content via the event in dragover (as it may be cross-document)
+// so we need shared state with the drag so we can draw previews of the drop in
+// the document.
+let draggedCode: string | undefined;
+let invertedChanges: ChangeSet | undefined;
+
+export const setDraggedCode = (value: string | undefined) => {
+  draggedCode = value;
+};
+
+export const snippetDropSupport = () => {
+  let lastPos: number | undefined;
+  const revertPreview = (view: EditorView) => {
+    if (invertedChanges) {
+      view.dispatch({
+        changes: invertedChanges,
+        annotations: [Transaction.addToHistory.of(false)],
+      });
+      invertedChanges = undefined;
+    }
+  };
+  return [
+    EditorView.domEventHandlers({
+      dragover(event, view) {
+        if (draggedCode) {
+          const pos = view.posAtCoords(event);
+          if (pos !== null && pos !== lastPos) {
+            revertPreview(view);
+            lastPos = pos;
+            const transaction = calculateChanges(view.state, draggedCode, pos);
+            // Stash the inverse for when we next update the doc — we'll "undo" first.
+            invertedChanges = transaction.changes.invert(
+              transaction.startState.doc
+            );
+            // Skip selection updates
+            view.dispatch({
```

Comment:
Would it help to combine the revertPreview transaction with this one?

Reply:
We could try this and see if it improves performance when throttling. I don't see the distinct states in the view so perhaps they're being collapsed by CM anyway.

Reply:
I think this is necessary for correctness. We're reading a position, reverting the preview, then using that position with the new state after the revert. So we either map it or collapse the changes into one transaction.

Reply:
I've mapped it for now so the logic for computing the changes doesn't have to account for the other edit. Resolving.